src/ng/compile.js

```diff
@@ -310,6 +316,39 @@
  *   they are waiting for their template to load asynchronously and their own compilation and linking has been
  *   suspended until that occurs.
  *
+ * ** $doCheck example **
+ *
+ * This example show how you might use `$doCheck` to customise the equality check of component inputs.
+ *
+ * <example name="doCheckExample" module="do-check-module">
+ *   <file name="index.html">
+ *     <div ng-init="items = []">
+ *       <button ng-click="items.push(items.length)">Add Item</button>
+ *       <button ng-click="items = []">Reset Items</button>
+ *       <pre>{{ items }}</pre>
+ *       <test items="items"></test>
+ *     </div>
+ *   </file>
+ *   <file name="app.js">
+ *      angular.module('do-check-module', [])
+ *        .component('test', {
+ *          bindings: { items: '<' },
+ *          template:
+ *            '<pre>{{ $ctrl.log | json }}</pre>',
+ *          controller: function() {
+ *            this.log = [];
+ *
+ *            this.$doCheck = function() {
+ *              if (this.items_ref !== this.items) { this.log.push('doCheck: items changed'); }
+ *              if (!angular.equals(this.items_clone, this.items)) { this.log.push('doCheck: items mutated'); }
+ *
+ *              this.items_clone = angular.copy(this.items);
+ *              this.items_ref = this.items;
+ *            };
```

Comment:
How about a simpler (and more lightweight example). We should better not encourage people to use `angular.copy` and `angular.equals` on potentially large arrays of potentially deep or large objects :smiley:

What happened to checking an internal field of a Date object? :smiley: